### PR TITLE
Build theia-docker in case no specific images were changed

### PR DESCRIPTION
**Pull request by dwjbosman**
Solution for #246

### CODE REVIEW

**Comment by vince-fugnitto** [comment]
@marcdumais-work @lmcbout 

I made the suggestion that the `theia-full-docker` image be built and tested when no images are explicitly updated during a PR. Should we however default to checking all images? (to be on the safe side). I see two main options:

1. always build all images if **no** images contain updates (might be expensive - we lose the functionality of only testing images that have changes).
2. test only `theia-full-docker` if **no** images contain updates, and restart CI jobs manually if we believe they should be tested (defeats the purpose of automating the CI job)

WDYT?

**Comment by lmcbout** [comment]
Solution 1: build all the times, every PR building everything: too expensive 

I think testing the theia-full-docker image should be enough, (Solution 2) . When a **PR** is merged, no need to test all images in the case where there is no changes affecting all images. 

We should have an option to compile selected images without any changes in that image upon a user request. May be we could pass a compilation flag that specify we want to build this/all image(s) with/without changes

May be another alternative is to build all **NEXT** images on a cron job at a specific time of the day every day !

What do you think?

**Comment by vince-fugnitto** [comment]
> We should have an option to compile selected images without any changes in that image upon a user request. May be we could pass a compilation flag that specify we want to build this/all image(s) with/without changes

This seems like overkill to me (and not commonly used). We can simply restart builds from the travis page manually (for example if I want to run `theia-cpp-docker`).

> May be another alternative is to build all **NEXT** images on a cron job at a specific time of the day every day !

I believe this is already the case.

**Comment by lmcbout** [comment]
>>We should have an option to compile selected images without any changes in that image upon a >>user request. May be we could pass a compilation flag that specify we want to build this/all >>image(s) with/without changes

>This seems like overkill to me (and not commonly used). We can simply restart builds from the travis >page manually (for example if I want to run theia-cpp-docker).

I think if you start the build from travis with no changes in the image,  with the latest changes done for issue [236,](https://github.com/theia-ide/theia-apps/pull/236) all images will skip building

>>May be another alternative is to build all NEXT images on a cron job at a specific time of the day >>every day !

>I believe this is already the case.

If this is the case , then perfect.

**Comment by vince-fugnitto** [comment]
> I think if you start the build from travis with no changes in the image, with the latest changes done for issue [236,](https://github.com/theia-ide/theia-apps/pull/236) all images will skip building

This PR builds the `theia-full-docker` in the following case. I believe it's enough, but just wanted feedback if we should build `theia-full-docker` or all images.

**Comment by lmcbout** [comment]
Good way to push your PR before mine  :)

**Comment by vince-fugnitto** [comment]
> Good way to push your PR before mine :)

I'm sorry, which PR?

**Comment by dwjbosman** [comment]
> I think if you start the build from travis with no changes in the image, with the latest changes done for issue [236,](https://github.com/theia-ide/theia-apps/pull/236) all images will skip building

#236 was adapted in such a way that it will build all images when the event type is not "pull_request". 

TRAVIS_EVENT_TYPE: 
- push -> all images tested **(is the correct behavior?)**
- pull_request -> 
  -  only specific image tested
  -  (#247) if the change was not for a specific image, theia-full-docker is tested
- api -> (eg. via travis  dashboard) -> all images tested
- cron -> all images tested

**Comment by dwjbosman** [comment]
> @dwjbosman I've spoken to @marcdumais-work and the `full` image might not be the best one to test since it's quite long and is the most fragile due to `gpg` keys.
> 
> Can we update the PR to instead use the [`theia-docker`](https://github.com/theia-ide/theia-apps/tree/master/theia-docker) image as the default case?

Changed it. Turned out there was a mistake in the regular expression as well.

File: check_changed.sh [hunk 1]
```diff
@@ -16,9 +16,39 @@ then
...
+        CHANGED_FILES_NON_DOCKER=$(git diff --name-status HEAD~1...HEAD .)
+        while read -r line; do
+            # the output of the git diff is of the form "M    theia-somename-docker/xyz"
+            if [[ $line =~ ^[[:space:]]*.[[:space:]]+theia(.)*-docker(.)+ ]]; then
+                # there was a change to an theia*-docker folder (not theia-docker)
+                # this will be picked up by another build
+                :
+            else
+                # there was a change in a non theia*-docker folder
+                # build theia-full-docker to be sure no utility scripts are broken
```
Comment (vince-fugnitto) [inline check_changed.sh]: The comment should be updated as well correct?
`# build theia-docker ...`

The same is true for the PR title.

**Comment by vince-fugnitto** [comment]
Thank you for making the updates @dwjbosman 👍

**Pull request merged**